### PR TITLE
Add Fusion offline authentication article

### LINKED ISSUE #349: Fusion: add offline authentication article
Splitted from vaadin/flow#10247

As we found out, the full `auth.ts` example code is too complex.

Add a separate article which explains handling authentication offline, using the `auth.ts` example snippets.

### PR DESCRIPTION
**Pull request by platosha**
Fixes #349

### CODE REVIEW

**Review by haijian-vaadin — CHANGES_REQUESTED** [review]
Could add a link to the new article in [Security Overview](https://vaadin.com/docs/latest/fusion/security/overview)

The link to spring-login in [Security Aspects of Offline Authentication Checks](https://vaadin.com/docs/latest/fusion/pwa/offline-authentication-check)

File: articles/fusion/security/authentication-offline.asciidoc [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+order: 110
+layout: page
+---
+
+= Offline Support for Authentication
+
+When building a Fusion application with offline support, take care of storing authentication in the browser to enable offline client-side access checking.
+
+The `localStorage` API is an easy way to store some data in the browser, that also works offline.
+Let us use that to store the authentication.
```
Comment (magi42) [inline articles/fusion/security/authentication-offline.asciidoc:12]: Could say: "This article describes how to use that to store authentication."

File: articles/fusion/security/authentication-offline.asciidoc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+title: Offline Support for Authentication
+order: 110
+layout: page
+---
+
+= Offline Support for Authentication
+
+When building a Fusion application with offline support, take care of storing authentication in the browser to enable offline client-side access checking.
+
+The `localStorage` API is an easy way to store some data in the browser, that also works offline.
```
Comment (magi42) [inline articles/fusion/security/authentication-offline.asciidoc:11]: * "some" is unnecessary
* No comma before "that".

File: articles/fusion/security/authentication-offline.asciidoc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+Let us use that to store the authentication.
+
+== Storing the Authentication for Offline
+
+It is a good idea to expire authentication after a certain time limit, but the data in `localStorage` does not expire automatically.
```
Comment (magi42) [inline articles/fusion/security/authentication-offline.asciidoc:16]: * "good idea" -> "good practice"
* "certain" is somewhat unnecessary

File: articles/fusion/security/authentication-offline.asciidoc [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+So you need to have a timestamp in the authentication data object itself.
+Add the timestamp property to the TypeScript definition for the authentication object:
+
+.`auth.ts`
+[source,typescript]
+----
+include::{root}/frontend/demo/fusion/authentication/auth.ts[tags=basic;!userinfo]
+----
+
+Also define the string key for `localStorage`, and the max age limit constant.
```
Comment (magi42) [inline articles/fusion/security/authentication-offline.asciidoc:26]: * "Also, "
* No comma before "and" (Oxford comma only used in lists of three or more)

File: articles/fusion/security/authentication-offline.asciidoc [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+When building a Fusion application with offline support, take care of storing authentication in the browser to enable offline client-side access checking.
+
+The `localStorage` API is an easy way to store some data in the browser, that also works offline.
+Let us use that to store the authentication.
+
+== Storing the Authentication for Offline
+
+It is a good idea to expire authentication after a certain time limit, but the data in `localStorage` does not expire automatically.
+So you need to have a timestamp in the authentication data object itself.
+Add the timestamp property to the TypeScript definition for the authentication object:
```
Comment (magi42) [inline articles/fusion/security/authentication-offline.asciidoc:18]: * "...as follows:"

File: articles/fusion/security/authentication-offline.asciidoc [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+Also define the string key for `localStorage`, and the max age limit constant.
+Examples in this article use a thirty days limit.
+
+.`auth.ts`
+[source,typescript]
+----
+include::{root}/frontend/demo/fusion/authentication/auth.ts[tags=storage-defs]
+----
+
+Then, in your login method, make the object with the timestamp upon successful login result, and save it using `localStorage.setItem()`:
```
Comment (magi42) [inline articles/fusion/security/authentication-offline.asciidoc:35]: * Difficult to understand where the object of the sentence is. Rephrase: "Then, in your login method, upon successful login, create the object with the timestamp, and save it..."

File: articles/fusion/security/authentication-offline.asciidoc [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+[source,typescript]
+----
+include::{root}/frontend/demo/fusion/authentication/auth.ts[tags=restore;!userinfo]
+----
+
+Remember to check the timestamp before using the loaded authentication.
+
+== Removing the Expired Authentication
+
+If the authentication is expired, or when the user logs out, remove the stored authentication from `localStorage`:
```
Comment (magi42) [inline articles/fusion/security/authentication-offline.asciidoc:59]: * "... as follows:"

File: articles/fusion/security/authentication-offline.asciidoc [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+.`auth.ts`
+[source,typescript]
+----
+include::{root}/frontend/demo/fusion/authentication/auth.ts[tags=logout]
+----
+
+== Complete Example With Offline Support
+
+Here is the complete client-side authentication code example with offline support:
```
Comment (magi42) [inline articles/fusion/security/authentication-offline.asciidoc:69]: * Avoid "here", rather use: "In the following"

File: articles/fusion/security/role-based-access.asciidoc [hunk 9]
```diff
@@ -63,43 +63,28 @@ include::{root}/frontend/demo/fusion/authentication/auth.ts[tags=isUserInRole]
...
-      await import('./views/products/products-view');
-      return undefined;
-    },
-  },
-  ...
-];
+include::{root}/frontend/demo/fusion/authentication/routes.ts[tags=ViewRoute]
+----
+
+Add a method for checking access for the given route by iterating `rolesAllowed` using `isUserInRole`:
```
Comment (magi42) [inline articles/fusion/security/role-based-access.asciidoc:74]: * "... as follows:"

File: articles/fusion/security/role-based-access.asciidoc [hunk 10]
```diff
@@ -63,43 +63,28 @@ include::{root}/frontend/demo/fusion/authentication/auth.ts[tags=isUserInRole]
...
+
+.`routes.ts`
+[source,typescript]
+----
+include::{root}/frontend/demo/fusion/authentication/routes.ts[tags=isAuthorizedViewRoute]
+----
+
+Then use the method added in the route action to redirect on unauthorized access:
```
Comment (magi42) [inline articles/fusion/security/role-based-access.asciidoc:82]: * "... as follows:"

**Review by magi42 — CHANGES_REQUESTED** [review]

**Review by haijian-vaadin — APPROVED** [review]

File: articles/fusion/pwa/offline-authentication-check.asciidoc [hunk 11]
```diff
@@ -9,15 +9,15 @@ layout: page
 Take the following security aspects into account when you do authentication checks while the user is offline
 
 == Authenticity
-Do you need to protect the local user's data on the device from whoever else may have access to the same device? If that is a requirement, then you will end up adding a `passcode` in addition to the regular online login (can be replaced with a fingerprint/face id auth through the WebAuthn API on the devices that support those). This is what many mobile bank apps do. 
-It is to encrypt any user data that is cached locally for offline use. 
+Do you need to protect the local user's data on the device from whoever else may have access to the same device? If that is a requirement, then you will end up adding a `passcode` in addition to the regular online login (can be replaced with a fingerprint/face id auth through the WebAuthn API on the devices that support those). This is what many mobile bank apps do.
```
Comment (magi42) [inline articles/fusion/pwa/offline-authentication-check.asciidoc:12]:
```suggestion
Do you need to protect the local user's data on the device from whoever else may have access to the same device?
If that is a requirement, you need to add a `passcode` in addition to the regular online login (or use fingerprint/face ID authentication through the WebAuthn API on the devices that support those).
This is what many mobile bank apps do.
```

File: articles/fusion/pwa/offline-authentication-check.asciidoc [hunk 12]
```diff
@@ -9,15 +9,15 @@ layout: page
...
+It is to encrypt any user data that is cached locally for offline use.
 If you do go that route, most likely you will end up having a helper library that you can call to check if the user is logged in or not.
 
 Arguably, not many apps would need to securely verify the user authenticity offline, and most can check if an online login happened in the not-so-distant past.
 
  - For that the simplest approach would be setting a local storage flag after a user has successfully logged in online (and clear it when the user logs out).
 During offline navigation, the user auth then would be checked from the local storage.
-The <<../security/spring-login#, Authentication with Spring Security>> article describes how to do that.
+The <<../security/authentication-offline.asciidoc#, Offline Support for Authentication>> article describes how to do that.
```
Comment (magi42) [inline articles/fusion/pwa/offline-authentication-check.asciidoc:20]:
```suggestion
The <<../security/authentication-offline#, Offline Support for Authentication>> article describes how to do that.
```

File: articles/fusion/security/role-based-access.asciidoc [hunk 13]
```diff
@@ -63,43 +63,28 @@ include::{root}/frontend/demo/fusion/authentication/auth.ts[tags=isUserInRole]
 
 == Routes With Access Control
 
-Use `isUserInRole` in the route action helper:
+Define an extended type `ViewRoute` that `rolesAllowed` string to enable specifying allowed roles on the view routes:
```
Comment (magi42) [inline articles/fusion/security/role-based-access.asciidoc:66]:
```suggestion
To enable specifying allowed roles on the view routes, define an extended type `ViewRoute` that has a `rolesAllowed` string as follows:
```

**Review by magi42 — CHANGES_REQUESTED** [review]

**Review by magi42 — APPROVED** [review]

**Review by magi42 — APPROVED** [review]

**Pull request merged**